pkg/iss/pbftsegmentchkp.go

```diff
@@ -0,0 +1,302 @@
+	// They contain the hashes of all Preprepare messages, as committed by the respective nodes.
+	// (A node sends a Done message when it commits everything in the segment.)
+	doneMessages map[t.NodeID]*isspbftpb.Done
+
+	// For each received Done message, stores the IDs of nodes that sent it.
+	doneMsgIndex map[string][]t.NodeID
+
+	// Once enough Done messages have been received,
+	// digests will contain the Preprepare digests for all sequence numbers in the segment.
+	digests map[t.SeqNr][]byte
+
+	// Once enough Done messages have been received,
+	// DoneNodes will contain IDs of nodes from which matching Done messages were received.
+	doneNodes []t.NodeID
+
+	// This flag is set once the retransmission of missing committed requests is requested.
+	// It serves preventing redundant retransmission requests when more than a quorum of Done messages are received.
+	catchingUp bool
+}
+
+// newPbftSegmentChkp returns a pointer to a new instance of pbftSegmentChkp
+func newPbftSegmentChkp() *pbftSegmentChkp {
+	return &pbftSegmentChkp{
+		doneMessages: make(map[t.NodeID]*isspbftpb.Done),
+		doneMsgIndex: make(map[string][]t.NodeID),
+	}
+}
+
+// Digests returns, for each sequence number of the associated segment, the digest of the committed batch
+// (more precisely, the digest of the corresponding Preprepare message).
+// If the information is not yet available (not enough Done messages have been received), Digests returns nil.
+func (chkp *pbftSegmentChkp) Digests() map[t.SeqNr][]byte {
+	return chkp.digests
+}
+
+// DoneNodes returns a list of IDs of Nodes from which a matching Done message has been received.
+// If a quorum of Done messages has not yet been received, DoneNodes returns nil.
+func (chkp *pbftSegmentChkp) DoneNodes() []t.NodeID {
+	return chkp.doneNodes
+}
+
+// Stable returns true if the instance-level checkpoint is stable,
+// i.e., if a strong quorum of matching Done messages has been received.
+// This ensures that at least a weak quorum of correct nodes has a local checkpoint
+// and thus evey correct node will be able to catch up.
+func (chkp *pbftSegmentChkp) Stable(numNodes int) bool {
+	for _, nodeIDs := range chkp.doneMsgIndex {
+		if len(nodeIDs) >= strongQuorum(numNodes) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// NodeDone registers a Done message received from a node.
+// Once NodeDone has been called with matching Done messages for a quorum of nodes,
+// the instance-level checkpoint will become stable.
+func (chkp *pbftSegmentChkp) NodeDone(nodeID t.NodeID, doneMsg *isspbftpb.Done, segment *segment) {
+
+	// Ignore duplicate Done messages.
+	if _, ok := chkp.doneMessages[nodeID]; ok {
+		return
+	}
+
+	// Store Done message
+	chkp.doneMessages[nodeID] = doneMsg
+	strKey := aggregateToString(doneMsg.Digests)
+	chkp.doneMsgIndex[strKey] = append(chkp.doneMsgIndex[strKey], nodeID)
+
+	// If a quorum of nodes has sent a Done message
+	if nodeIDs := chkp.consistent(len(segment.Membership)); nodeIDs != nil {
```

Comment:
Why do we go over the whole `chkp.doneMsgIndex` inside `chkp.consistent` if we know that only `chkp.doneMsgIndex[strKey]` has changed?

Reply:
Good point :-) Fixed.